### PR TITLE
Fix: Optional pre-type won't cause a crash anymore

### LINKED ISSUE #5369: Crash in new resolver
The following program crashes the resolver

```dafny
ghost function f2(items: set<nat>): (r:nat)
  requires |items| > 0 {
    var item :| item in items;
    item
}

function f(items: set<nat>) : (r:nat)
requires |items| > 0 {
    //assume exists item :: item in items && forall other <- items :: item == other || item < other;
    //assert exists item Smallest()
    var item :| IsSmallest(items, item);
    item
}

predicate IsSmallest(s: set<nat>, item: nat) 
  requires item in s {
  m in s && forall y: nat :: y in s ==> m <= y
}

lemma Smallest(s: set<nat>) returns (m: nat)
  requires s != {}
  ensures m in s && IsSmallest(s, m)
  decreases s
{
  m :| m in s;
  if y: nat :| y in s && y < m {
    ghost var s' := s - {m};
    assert y in s';
    m := ThereIsASmallest(s');
  }
}

function smallest(items: set<nat>): (r: nat)
  requires |items| > 0

method m(items: set<nat>) returns (r:nat)
requires |items| > 0 {
    var item :| item in items;
    return item;
}
```

results in:

```
Dafny encountered an internal error. Please report it at <https://github.com/dafny-lang/dafny/issues>.
System.InvalidCastException: Unable to cast object of type 'Microsoft.Dafny.UnusedPreType' to type 'Microsoft.Dafny.DPreType'.
   at Microsoft.Dafny.PreTypeConstraints.ConfirmConstraint(CommonConfirmationBag check, PreType preType, DPreType auxPreType)
   at Microsoft.Dafny.PreTypeConstraints.ConfirmTypeConstraints()
   at Microsoft.Dafny.PreTypeConstraints.SolveAllTypeConstraints(String printableContext)
   at Microsoft.Dafny.PreTypeResolver.ResolveFunction(Function f)
   at Microsoft.Dafny.PreTypeResolver.ResolveMember(MemberDecl member)
   at Microsoft.Dafny.PreTypeResolver.ResolveDeclarationBody(Declaration d)
   at Microsoft.Dafny.PreTypeResolver.ResolveDeclarations(List`1 declarations, ModuleResolver resolver, Boolean firstPhaseOnly)
   at Microsoft.Dafny.ModuleResolver.ResolveTopLevelDecls_Core(List`1 declarations, Graph`1 datatypeDependencies, Graph`1 codatatypeDependencies, String moduleDescription, Boolean isAnExport)
   at Microsoft.Dafny.ModuleDefinition.Resolve(ModuleSignature sig, ModuleResolver resolver, String exportSetName)
   at Microsoft.Dafny.LiteralModuleDecl.Resolve(ModuleResolver resolver, CompilationData compilation)
   at Microsoft.Dafny.ModuleResolver.ResolveModuleDeclaration(CompilationData compilation, ModuleDecl decl)
   at Microsoft.Dafny.LanguageServer.Language.CachingResolver.ResolveModuleDeclaration(CompilationData compilation, ModuleDecl decl)
   at Microsoft.Dafny.ProgramResolver.Resolve(CancellationToken cancellationToken)
   at Microsoft.Dafny.LanguageServer.Language.CachingResolver.<>c__DisplayClass5_0.<Resolve>b__0()
   at Microsoft.Dafny.LanguageServer.Language.CachingResolver.Resolve(CancellationToken cancellationToken)
   at Microsoft.Dafny.LanguageServer.Language.Symbols.DafnyLangSymbolResolver.RunDafnyResolver(Compilation compilation, Program program, CancellationToken cancellationToken)
   at Microsoft.Dafny.LanguageServer.Language.Symbols.DafnyLangSymbolResolver.ResolveSymbols(Compilation compilation, Program program, CancellationToken cancellationToken)
   at Microsoft.Dafny.TextDocumentLoader.ResolveInternal(Compilation compilation, Program program, CancellationToken cancellationToken)
   at Microsoft.Dafny.TextDocumentLoader.<>c__DisplayClass5_0.<<ResolveAsync>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.Dafny.TextDocumentLoader.ResolveAsync(Compilation compilation, Program program, CancellationToken cancellationToken)
   at Microsoft.Dafny.Compilation.ResolveAsync()
```

### PR DESCRIPTION
**Pull request by MikaelMayer**
This PR fixes #5369
I added the corresponding test.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/Resolver/PreType/PreTypeConstraints.cs [hunk 1]
```diff
@@ -583,6 +583,8 @@ public enum CommonConfirmationBag {
       preType = preType.Normalize();
       if (preType is PreTypeProxy) {
         return false;
+      } else if (preType is UnusedPreType) {
```
Comment (keyboardDrummer) [inline Source/DafnyCore/Resolver/PreType/PreTypeConstraints.cs:586]: Can you add a comment to the class `UnusedPreType` ?

File: Source/DafnyCore/Resolver/PreType/PreType.cs [hunk 2]
```diff
@@ -455,6 +455,9 @@ public class PreTypePlaceholderModule : PreTypePlaceholder {
   public class PreTypePlaceholderType : PreTypePlaceholder {
   }
 
+  /// Currently used exclusively for assigning a pre-type to MemberSelect expressions, such as "obj.method",
+  /// which is not considered an expression. This indicates that resolution has occurred,
```
Comment (keyboardDrummer) [inline Source/DafnyCore/Resolver/PreType/PreType.cs:459]: What do you mean by `MemberSelect` not being considered an expression? It inherits from `Expression`

Comment (MikaelMayer) [reply Source/DafnyCore/Resolver/PreType/PreType.cs:459]: It weirdly is an Expression, but not a true Dafny expression: you cannot use it in a regular Dafny expression. Consider the following:

```
class A {
  method Test() {
  }
}
method Main() {
  var a := new A();
  var f := a.Test; // Error, expected a Dafny expression on the RHS
  f();
}
```

Comment (keyboardDrummer) [reply Source/DafnyCore/Resolver/PreType/PreType.cs:459]: It is also used for fields, in which case it is an expression, right?

File: Source/DafnyCore/Resolver/PreType/PreTypeConstraints.cs [hunk 3]
```diff
@@ -583,6 +583,8 @@ public enum CommonConfirmationBag {
       preType = preType.Normalize();
```
Comment (keyboardDrummer) [inline Source/DafnyCore/Resolver/PreType/PreTypeConstraints.cs:583]: Can you add a comment to `ConfirmConstraint`, or update its signature (parameter names) to make it clear what it does? I don't understand the meaning of the parameters or return value by looking at just its signature and comment.

File: Source/DafnyCore/Resolver/PreType/PreType.cs (hunk 2 above)
Comment (keyboardDrummer) [reply Source/DafnyCore/Resolver/PreType/PreType.cs:459]: It seems nice to conceptually consider `MemberSelectExpr` that references a method, still an expression, even if we don't allow it in an expression context. If we do that, then I would expect to assign either a function type to it or a custom type `GenericCallableType`. Maybe we can rename `UnusedPreType` to `MethodType`

File: Source/DafnyCore/Resolver/PreType/PreTypeConstraints.cs (hunk 3 above)
Comment (MikaelMayer) [reply Source/DafnyCore/Resolver/PreType/PreTypeConstraints.cs:583]: I can't, as I don't understand the code at a level high enough to do that. I don't have a mental model of what ConfirmConstraint does, nor about what code context it was called with. I'll leave it for @RustanLeino  to comment on ti.

File: Source/DafnyCore/Resolver/PreType/PreType.cs (hunk 2 above)
Comment (MikaelMayer) [reply Source/DafnyCore/Resolver/PreType/PreType.cs:459]: Yes, it's used for fields and in that case it's indeed an expression.

I did the renaming, hoping it won't break any future development.

**Comment by keyboardDrummer** [comment]
I don't think this is the right fix. Looking at the input program:

```
predicate IsSmallest(s: set<nat>, item: nat) 
  requires item in s {
  m in s && forall y: nat :: y in s ==> m <= y //<-- the problem is that this resolves to the method `m` but it should not since it should only resolve to something that is an expression. If you let it resolve to the method `m`, then you get weird behavior such as that it tries to resolve `m <= y` with `m` having this unused MethodPreType.
}

method m(items: set<nat>) returns (r:nat)
requires |items| > 0 {
    var item :| item in items;
    return item;
}
```

**Review by keyboardDrummer — APPROVED** [review]

**Comment by MikaelMayer** [comment]
> I don't think this is the right fix. Looking at the input program:

Thank you very much for this more precise fix ! Looks great.

**Pull request merged**